manager_janitor.go

```diff
@@ -335,6 +335,105 @@ func (mgr *Manager) pindexesRestart(
 	return errs
 }
 
+func (mgr *Manager) activatePIndex(pi *pindexRestartReq) (*PIndex, error) {
+	err := mgr.stopPIndex(pi.pindex, false)
+	if err != nil {
+		log.Printf("janitor: error closing pindex %s", pi.pindex.Name)
+		return nil, err
+	}
+	// rename the pindex folder and name as per the new plan
+	newPath := mgr.PIndexPath(pi.planPIndexName)
+	if newPath != pi.pindex.Path {
+		log.Printf("janitor: proceeding to rename %s to %s", pi.pindex.Path, newPath)
+		err := os.Rename(pi.pindex.Path, newPath)
+		if err != nil {
+			cleanDir(pi.pindex.Path)
+			cleanDir(newPath)
+			return nil, fmt.Errorf("janitor: restartPIndex"+
+				" updating pindex: %s path: %s failed, err: %v",
+				pi.pindex.Name, newPath, err)
+		}
+		_, err2 := os.Stat(pi.pindex.Path)
+		if err2 == nil { // old pindex dir still exists.
+			err := os.RemoveAll(pi.pindex.Path) // quick fix - deleting it.
+			if err != nil {
+				return nil, fmt.Errorf("janitor: error removing path %s: %s",
+					pi.pindex.Path, err.Error())
+			}
+		}
+		_, err2 = os.Stat(newPath)
+		if os.IsNotExist(err2) {
+			return nil, fmt.Errorf("janitor: error renaming to %s: %s",
+				newPath, err.Error())
+		}
+	}
+	pi2 := pi.pindex.Clone() //pi2 is the new pindex.
+	pi2.Name = pi.planPIndexName
+	pi2.Path = newPath
+
+	_ = mgr.unregisterPIndex(pi.pindex.Name, pi.pindex)
+	// unreg the old pindex here, not in hibernatePIndexUtil, since after this, this pidx won't be used anymore
+	// in cold, it might need to be used again, so just hibernatePIndexUtil, not unreg.
+
+	// persist PINDEX_META only if manager's dataDir is set
+	if len(mgr.dataDir) > 0 {
+		// update the new indexdef param changes
+		buf, err := json.Marshal(pi2)
+		if err != nil {
+			cleanDir(newPath)
+			return nil, fmt.Errorf("janitor: tempRestartPIndex"+
+				" Marshal pindex: %s, err: %v", pi2.Name, err)
+
+		}
+		err = ioutil.WriteFile(pi2.Path+string(os.PathSeparator)+
+			PINDEX_META_FILENAME, buf, 0600)
+		if err != nil {
+			cleanDir(pi2.Path)
+			return nil, fmt.Errorf("janitor: tempRestartPIndex could not save "+
+				"PINDEX_META_FILENAME,"+" path: %s, err: %v", pi2.Path, err)
+		}
+	}
+
+	// open the pindex and register
+	// since this is done for hot indexes, an open pidx intuitively makes sense since
+	// a hot pidx is one that is queried freq and so it should be open
+	pindex, err := OpenPIndex(mgr, pi2.Path)
+	if err != nil {
+		log.Printf("error in opening path %s", pi2.Path)
+		cleanDir(pi.pindex.Path)
+		return nil, fmt.Errorf("janitor: tempRestartPIndex could not open "+
+			" pindex path: %s, err: %v", pi2.Path, err)
+	}
+	err = mgr.registerPIndex(pindex)
+	if err != nil {
+		log.Printf("error in registering pindex %s", pindex.Name)
+		cleanDir(pindex.Path)
+		return nil, fmt.Errorf("janitor: tempRestartPIndex failed to "+
+			"register pindex: %s, err: %v", pindex.Name, err)
+	}
+	return pindex, nil
```

Comment:
activatePIndex() already does a ton of duplicate things and then you again calling a hibernatePIndexUtil() for further actions.
Please reconcile all these methods and avoid patterns like "action - revert" / "do-undo" for various operations.

Reply:
Thanks.
To try and avoid duplicate work, I have added an additional 'complete' parameter for incomplete and complete activations.
If an activation is going to be immediately followed by a hibernation(i.e. cold index), the index can be incompletely activated.
If it is not going to be followed by a hibernation(i.e hot/warm indexes), it is completely activated.

Reply:
There are still a couple of things I need to finalise here(source partitions, test with endpoint) but this is the gist of it.